tests/Unit/Products/ProductImageTest.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+namespace Tests\Unit\Products;
+
+use Illuminate\Contracts\Filesystem\Factory as FilesystemContract;
+use Illuminate\Database\QueryException;
+use Illuminate\Http\UploadedFile;
+use IndieHD\Velkart\Base\Traits\UploadsFiles;
+use IndieHD\Velkart\ProductImage\Contracts\ProductImageRepositoryContract;
+use IndieHD\Velkart\Tests\TestCase;
+
+class ProductImageTest extends TestCase
+{
+    use UploadsFiles;
+
+    protected $repo;
+
+    protected $filesystem;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->repo = resolve(ProductImageRepositoryContract::class);
+        $this->filesystem = resolve(FilesystemContract::class);
+    }
+
+    private function createProductImage($params = null): object
+    {
+        if ($params === null)
+            $params = factory($this->repo->modelClass())->make()->toArray();
+
+        return $this->repo->create($params);
+    }
+
+    /** @test */
+    public function itCanCreateAProductImage()
+    {
+        $this->assertDatabaseHas('product_images', ['src' => $this->createProductImage()->src]);
```

Comment:
I agree with the hard coded concern ... however I can't imagine why we would change the table name as we follow a fairly strict name convention. I feel like when using instanceOf for every assert it's hard to really understand what's being tested. Because if I read that line, it can be a bit confusing how we are actually validating the creation of said resource based on an assert instanceOf ... 

Not really sure what the right answer is here ... 

The other option is to contract the repository and model to have a method that is like

```php
$this->repo->table();
```

...

Reply:
Even if we decide that `instanceOf()` is ambiguous, I still think this is better:

```php
$this->assertNotNull($this->repo->find($this->createProductImage()->id));
```

*We* may not ever change the table name, but that doesn't mean an end-user won't want to.

It could be as simple as the end-user wanting to add a prefix to all tables.

Yes, acquiring the table name dynamically is another option, but why bother when the above is more abstract and robust? What's the downside?

Reply:
The only downside is readability imo ...

I mean it is common practice for laravel packages to include configuration files for exactly this reason. But that would require dynamic migrations also and im not real familiar with that and its probably more complicated then what its worth.

I'll agree to move forward with an alternative method of testing persistence to the database.

Reply:
Actually, I'm content to leave this as-is, for two reasons:

1. On second thought, end-users won't need to run these tests against their own modifications (they would be running their own tests against their specific implementations), so that's not a valid concern, after all.
2. Using `$this->repo->find()` is worse than using `databaseHas()` because that's *our* code, not the test suite's. In other words, a test could fail because there's a problem with our repo's `find()` method when in fact the data was persisted successfully.

As long as there is no potential for database-related side-effects between tests, i.e., the changes from every test are always rolled-back before the next test is run, this should work reliably.